File: Software3DRenderer/SWRRenderContext.m
//
//  SWRRenderContext.m
//  Software3DRenderer
//
//  Created by Denys Kunytskyi on 8/19/15.
//  Copyright (c) 2015 Denys Kunytskyi. All rights reserved.
//

#import "SWRRenderContext.h"
@interface SWRRenderContext()
{
    int* _scanBuffer;
}
@end

@implementation SWRRenderContext

-(instancetype) initWithWidth:(int) w height:(int) h
{
    self=[super initWithWidth:w height:h colorSpace:SWRBITMAP_RGBA];
    if(self)
    {
        _scanBuffer=malloc(sizeof(int)*h*2);
    }
    return self;
}

-(void) dealloc
{
    free(_scanBuffer);
}

-(void) drawScanBufferAtY:(int) yCoord xMin:(int)xmin xMax:(int) xmax
{
    if(yCoord>=[self getSize].height)
        return;
    
    _scanBuffer[yCoord*2  ]=xmin;
    _scanBuffer[yCoord*2+1]=xmax;
}

-(void) fillShapeWithYMin:(int) ymin yMax:(int) ymax
{
    for (int j=ymin; j<ymax; j++)
    {
        int xMin=_scanBuffer[j*2  ];
        int xMax=_scanBuffer[j*2+1];
        for (int i=xMin;i<xMax;i++)
        {
            [self drawPixelAtpX:i pY:j withA:255 B:255 G:255 R:255];
        }
    }
}


@end

File: Software3DRenderer/SWR/SWRRenderContext.cpp
//
//  SWRRenderContext.cpp
//  Software3DRenderer
//
//  Created by Denys Kunytskyi on 8/20/15.
//  Copyright (c) 2015 Denys Kunytskyi. All rights reserved.
//

#include "SWRRenderContext.h"


SWRRenderContext::SWRRenderContext(int w, int h):SWRBitmap(w,h,SWRBITMAP_RGBA)
{
    _scanBuffer=new int[h*2];
}

SWRRenderContext::~SWRRenderContext()
{
    delete [] _scanBuffer;
}

void SWRRenderContext::DrawScanBuffer(int yCoord, int xMin, int xMax)
{
    if(yCoord>=_height)
        return;
    
    _scanBuffer[yCoord*2  ]=xMin;
    _scanBuffer[yCoord*2+1]=xMax;
}

void SWRRenderContext::FillShape(int yMin, int yMax)
{
    for (int j=yMin; j<yMax; j++)
    {
        int xMin=_scanBuffer[j*2  ];
        int xMax=_scanBuffer[j*2+1];
        for (int i=xMin;i<xMax;i++)
        {
            DrawPixel(i, j, 255, 255, 255, 255);
        }
    }
}
